// @azure/monitor-query/dist-esm/src/generated/metrics/src/models/index.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
/** Known values of {@link ApiVersion201801} that the service accepts. */
export var KnownApiVersion201801;
(function (KnownApiVersion201801) {
    /** Api Version '2018-01-01' */
    KnownApiVersion201801["TwoThousandEighteen0101"] = "2018-01-01";
})(KnownApiVersion201801 || (KnownApiVersion201801 = {}));
/** Known values of {@link MetricUnit} that the service accepts. */
export var KnownMetricUnit;
(function (KnownMetricUnit) {
    KnownMetricUnit["Count"] = "Count";
    KnownMetricUnit["Bytes"] = "Bytes";
    KnownMetricUnit["Seconds"] = "Seconds";
    KnownMetricUnit["CountPerSecond"] = "CountPerSecond";
    KnownMetricUnit["BytesPerSecond"] = "BytesPerSecond";
    KnownMetricUnit["Percent"] = "Percent";
    KnownMetricUnit["MilliSeconds"] = "MilliSeconds";
    KnownMetricUnit["ByteSeconds"] = "ByteSeconds";
    KnownMetricUnit["Unspecified"] = "Unspecified";
    KnownMetricUnit["Cores"] = "Cores";
    KnownMetricUnit["MilliCores"] = "MilliCores";
    KnownMetricUnit["NanoCores"] = "NanoCores";
    KnownMetricUnit["BitsPerSecond"] = "BitsPerSecond";
})(KnownMetricUnit || (KnownMetricUnit = {}));
