// @azure/monitor-query/dist-esm/src/generated/metricsnamespaces/src/models/index.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
/** Known values of {@link ApiVersion20171201Preview} that the service accepts. */
export var KnownApiVersion20171201Preview;
(function (KnownApiVersion20171201Preview) {
    /** Api Version '2017-12-01-preview' */
    KnownApiVersion20171201Preview["TwoThousandSeventeen1201Preview"] = "2017-12-01-preview";
})(KnownApiVersion20171201Preview || (KnownApiVersion20171201Preview = {}));
/** Known values of {@link NamespaceClassification} that the service accepts. */
export var KnownNamespaceClassification;
(function (KnownNamespaceClassification) {
    KnownNamespaceClassification["Platform"] = "Platform";
    KnownNamespaceClassification["Custom"] = "Custom";
    KnownNamespaceClassification["Qos"] = "Qos";
})(KnownNamespaceClassification || (KnownNamespaceClassification = {}));
